### ISSUE TITLE
Got no improvement with GAT, am I used a wrong way to process data?

### ISSUE THREAD

**Issue by w21180239** [pyg-team/pytorch_geometric#555]
## ❓ Questions & Help

Hello, I'm doing a subway traffic problem with GAT using PyG, but I found the result is poorly improved (even sometimes drop) compared with no edge graph.

Here is my code, please help me find out mistakes.

## preprocess step:
```
edge_index = list(G.edges())
edge_index = torch.tensor(edge_index).t().contiguous()
edge_index = edge_index - edge_index.min()

total = len(x_list)
for x, y in zip(x_list, y_list):
    data = Data(edge_index=edge_index, x=x, y=y)
    if i > int(total * 0.8):
        val_list.append(data)
    else:
        train_list.append(data)
    i += 1

torch.save(train_list, 'subway_data_train.npz')
torch.save(val_list, 'subway_data_val.npz')
```
## load data step:
```
train_data_list = torch.load('subway_data_train.npz')
val_data_list = torch.load('subway_data_val.npz')

train_loader = DataLoader(
    train_data_list, batch_size=batchsize, shuffle=True)

val_loader = DataLoader(
    val_data_list, batch_size=batchsize, shuffle=True)
```
## model:
```
class Net(torch.nn.Module):
    def __init__(self, FEA, OUT):
        super(Net, self).__init__()
        self.mlp = MLP([64, 128, 256, 128, 64])
        self.heads = nn.ModuleList([MLP([64, 32,32, 1]) for i in range(3)])

        if mmodel == 'GAT':
            self.conv1 = GATConv(FEA, 12, heads=12)
            self.conv2 = GATConv(
                12 * 12, 12, heads=12)
            self.conv3 = GATConv(
                12 * 12, 12, heads=12)
            self.conv4 = GATConv(
                12 * 12, 8, heads=8)

    def forward(self, x, edge_index):
        if mmodel == 'GAT':
            x = F.dropout(x, p=drop_rate, training=self.training)
            x = F.elu(self.conv1(x, edge_index))
            x = F.dropout(x, p=drop_rate, training=self.training)
            x = F.elu(self.conv4(x, edge_index))
            x = self.mlp(x)
            out_list = [head(x) for head in self.heads]
            x = torch.cat(out_list, 1)
```
## train step:
```
    for data in loader:
        optimizer.zero_grad()
        out = model(data.x.to(device), data.edge_index.to(device))
        loss = F.mse_loss(out, data.y.to(device), size_average=False)
        loss.backward()
        optimizer.step()
```

**Comment by rusty1s** [comment]
This is hard to tell. Your code looks fine to me. Does the model train at all or does it just poorly generalize? You may want to reduce your dropouts. In addition,
```python
edge_index = edge_index - edge_index.min()
```
looks suspicious to me (in case `edge_index.min()` is greater than `0`).

**Comment by w21180239** [comment]
> This is hard to tell. Your code looks fine to me. Does the model train at all or does it just poorly generalize? You may want to reduce your dropouts. In addition,
> 
> ```python
> edge_index = edge_index - edge_index.min()
> ```
> 
> looks suspicious to me (in case `edge_index.min()` is greater than `0`).

I'm pretty sure that `edge_index = edge_index - edge_index.min()` has no influence on the result cause I have tested it.

The model performs almost the same in every aspect (including train_loss, val_loss, etc) no matter the edge exists or not. It make me suspect that the GAT doesn't work at all in my code, it can't gather little meaningful information from neighbors. Is there any suggestion from you about how to find out the GATConv is working or not?

**Comment by rusty1s** [comment]
Just some random thoughts:
1. You could look at the gradients of the `GATConv` parameters to verify that the gradients are not all close to zero.
2. You could look at the quality of the generated attention scores. Are the mostly sharp or totally fuzzy?
3. You could swap out GNN operators (*e.g.*, `GCNConv`) to see if the problem remains.
4. You could learn a weighting between `GATConv` output and linear regression output to see which way the model prefers.

**Comment by w21180239** [comment]
Thank you very much for you reply, I have tried to change Conv layer (AGNN, ARMA, Genie) and the performance doesn't improve much, but I will try other ways you mentioned.

**Comment by kunwuz** [comment]
> Thank you very much for you reply, I have tried to change Conv layer (AGNN, ARMA, Genie) and the performance doesn't improve much, but I will try other ways you mentioned.

HI, have you found the effective way? Same puzzlement now

**Comment by wasdWarrior** [comment]
> ## ❓ Questions & Help
> Hello, I'm doing a subway traffic problem with GAT using PyG, but I found the result is poorly improved (even sometimes drop) compared with no edge graph.
> 
> Here is my code, please help me find out mistakes.
> 
> ## preprocess step:
> ```
> edge_index = list(G.edges())
> edge_index = torch.tensor(edge_index).t().contiguous()
> edge_index = edge_index - edge_index.min()
> 
> total = len(x_list)
> for x, y in zip(x_list, y_list):
>     data = Data(edge_index=edge_index, x=x, y=y)
>     if i > int(total * 0.8):
>         val_list.append(data)
>     else:
>         train_list.append(data)
>     i += 1
> 
> torch.save(train_list, 'subway_data_train.npz')
> torch.save(val_list, 'subway_data_val.npz')
> ```
> 
> ## load data step:
> ```
> train_data_list = torch.load('subway_data_train.npz')
> val_data_list = torch.load('subway_data_val.npz')
> 
> train_loader = DataLoader(
>     train_data_list, batch_size=batchsize, shuffle=True)
> 
> val_loader = DataLoader(
>     val_data_list, batch_size=batchsize, shuffle=True)
> ```
> 
> ## model:
> ```
> class Net(torch.nn.Module):
>     def __init__(self, FEA, OUT):
>         super(Net, self).__init__()
>         self.mlp = MLP([64, 128, 256, 128, 64])
>         self.heads = nn.ModuleList([MLP([64, 32,32, 1]) for i in range(3)])
> 
>         if mmodel == 'GAT':
>             self.conv1 = GATConv(FEA, 12, heads=12)
>             self.conv2 = GATConv(
>                 12 * 12, 12, heads=12)
>             self.conv3 = GATConv(
>                 12 * 12, 12, heads=12)
>             self.conv4 = GATConv(
>                 12 * 12, 8, heads=8)
> 
>     def forward(self, x, edge_index):
>         if mmodel == 'GAT':
>             x = F.dropout(x, p=drop_rate, training=self.training)
>             x = F.elu(self.conv1(x, edge_index))
>             x = F.dropout(x, p=drop_rate, training=self.training)
>             x = F.elu(self.conv4(x, edge_index))
>             x = self.mlp(x)
>             out_list = [head(x) for head in self.heads]
>             x = torch.cat(out_list, 1)
> ```
> 
> ## train step:
> ```
>     for data in loader:
>         optimizer.zero_grad()
>         out = model(data.x.to(device), data.edge_index.to(device))
>         loss = F.mse_loss(out, data.y.to(device), size_average=False)
>         loss.backward()
>         optimizer.step()
> ```

Hi, i have a similar bug like u, could you upload your entire code?

**Comment by rusty1s** [comment]
Closing this issue now. Feel free to re-open if there are still concerns about a bug in PyG.
